### PR TITLE
Add module index to sidebar

### PR DESCRIPTION
**Pull request by bradcray**
This adds a module index to our sidebar.    I'd seen mentions of this
index, but couldn't find it when I was adding other indices last month.
Stumbled across it today, so added it to the sidebar and renamed
things in the sidebar a bit to reflect it as well.

While here, I did a cleanup of the one-line summaries of each of the
module files to make the index cleaner than the (unlinked) file that
had been available through the website.  I also added some special-
case code to the compiler to hide internal modules from the index,
which also had the impact of making them not cross-referenceable.
I think this is appropriate, though, since the internal module names
aren't meant to be things that the user knows about or can refer to.
So then I had to update links to internal modules to link to other
things instead (a pre-existing type or section name, on in the case
of bytes and string, a new type annotation).

### CODE REVIEW

**Comment by bradcray** [comment]
@arezaii : This one's for you.  Would you mind reviewing?

**Review by arezaii — APPROVED** [review]
This is a nice addition to the menu. I think (in the future) we may want to look into what generates the text in the second column, as some have a nice short about message, some say  `.. warning::`, and others are blank or just have the word `Summary`. I don't think that is necessary for this change though.

**Comment by bradcray** [comment]
> we may want to look into what generates the text in the second column

Oh good catch (I obviously didn't look very hard at the result upon finding it).  I think this is just the first line of each module.  I'd been wondering about this recently, so may just take care of it before merging this PR while we're thinking of it.

**Comment by arezaii** [comment]
These changes make it much nicer to identify interesting modules. 

I noticed that there are only a few that don't have descriptions (or have odd descriptions) now: 

- BlockCycDim
- BlockCycDist
- BlockDim
- BlockDist
- ChapelArray
- ChapelDomain
- ChapelLocale (says "None")
- CyclicDist
- DimensionalDist2D
- LayoutCS
- OwnedObject (`owned` (along with :record:`~SharedObject.shared`) manage the)
- PrivateDist
- ReplicatedDim
- ReplicatedDist
- StencilDist

It may be that we don't really want those in this list and that's why they don't have descriptions, what do you think?

**Comment by bradcray** [comment]
I'm not done with the PR yet.  The internal modules I'm working on squashing from the index altogether.  The distributions don't have entries because they don't have a module keyword, but I'm not going to take that on in this PR (because I think a reasonable entry without a summary string is more acceptable than an unreasonable entry or a bad summary string).  Stay tuned.

**Comment by bradcray** [comment]
@arezaii : OK, this is ready for re-review now.

**Comment by vasslitvinov** [comment]
I prefer to see the two indices at the bottom of the "Writing Chapel Programs" section, rather than in their own section. That way I am more likely to notice their existence.

From another angle, the module index fits squarely under "Writing Chapel Programs". The "complete docs index" contains only items relevant to "Writing Chapel Programs", too, at least I did not see anything to the contrary (ex. CHPL_ environment variables).

... except for `dyno` library symbols `chpl::uast::XXX` and similar. Which perhaps should be in a different index instead, perhaps linked at `https://chapel-lang.org/docs/developer/compiler-internals/index.html`

This latter does not affect this PR, however would be a nice housecleaning item.

**Comment by vasslitvinov** [comment]
I noticed that the methods marked as "built-in procedure" are alphabetized by the receiver type in the "complete docs index", whereas all other methods are alphabetized by the method name. Examples include `tuple.indices()`, `tuple.size()`, and methods on `range` and `enum`. Not sure how that relates to this PR, however it would be another nice housecleaning item to fix their ordering. While at that, call them "methods" rather than "procedures".

**Comment by bradcray** [comment]
> I prefer to see the two indices at the bottom of the "Writing Chapel Programs" section, rather than in their own section. That way I am more likely to notice their existence.

When I added the original index to the sidebar, I didn't have much control over where it landed.  It seems ridiculous—and I could've easily been missing something—but based on web searches, it's pretty hard to inject things into the sidebar, since it's created by Sphinx on our behalf.  Another way to make it more noticeable would be to reduce the number of things on the sidebar, which currently seems a bit high.  For example, built-in types and functions should be going away, package modules and mason modules should become one thing over time, we could move all modules into a single page that has a submenu when clicked on, etc.

> The "complete docs index" contains only items relevant to "Writing Chapel Programs", too

It's a complete index, it's just that there aren't as many index entries created (neither by us, nor automatically) for the other sections.  As an example, see https://chapel-lang.org/docs/technotes/dsi.html#GlobalDomain

> Which perhaps should be in a different index instead

I agree that having distinct indexes for different things would be useful, but don't know how to accomplish that offhand.

> I noticed that the methods marked as "built-in procedure" are alphabetized by the receiver type in the "complete docs index", whereas all other methods are alphabetized by the method name.

I'm not sure what it would take to fix this, and don't think it relates to this PR.  I'm not trying to do everyone's housecleaning for them here, just trying to enable the module index.

>  we could consider switching the "complete docs index" from two columns to one.

This also seems distinct from this PR and isn't something I happen to know how to do.

File: compiler/AST/ModuleSymbol.cpp
Comment (bradcray) [inline compiler/AST/ModuleSymbol.cpp:363]: @lydia-duncan : This is the code that relates to my "Can chpldoc know that a module is internal" question the other day.  Would you give it a review and let me know whether you're OK with it going in?  (it's a bit fragile, but the only real consequence of getting the logic wrong is that a module will either appear correctly or incorrectly in our module index).

File: doc/rst/language/spec/classes.rst
Comment (bradcray) [inline doc/rst/language/spec/classes.rst:211]: @mppf: Would you be willing to review the changes to this part of the spec?  Essentially, the changes here were:
* we don't want SharedObject/OwnedObject to show up in the module index
* squashing them from the index also squashes the ability to cross-reference them
* however, that seems appropriate given that we really want them to be part of the spec / not known to the user as modules anyway
As a result, I anchored these to the record type declarations instead.  Other tricks could be used to link to other parts of the document (though I'm imagining that this may all get rejiggered once we bring these docs into the spec anyway...).

File: modules/internal/String.chpl
Comment (bradcray) [inline modules/internal/String.chpl:1801]: @e-kayrakli : If you'd be willing to review the changes I've made to the Bytes and String modules, I'd appreciate it.  I believe they reflect where we ended up in our discussion of different approaches, but would appreciate if you'd double-check me.

File: doc/rst/language/spec/classes.rst
Comment (mppf) [reply doc/rst/language/spec/classes.rst:211]: LGTM

**Review by e-kayrakli — APPROVED** [review]
String/Bytes changes look good.

File: modules/packages/Futures.chpl
Comment (bradcray) [inline modules/packages/Futures.chpl:23]: What is this?

Comment (mppf) [reply modules/packages/Futures.chpl:23]: u+2014 is an m-dash, FWIW

**Comment by bradcray** [comment]
I'd like to request having folks review the one-line descriptions that I added for modules that they added or shepherded in, in part to share the reviewing load, and in part to create more group awareness of the need to get these first lines of module documentation correct going forward.  Note that the descriptions don't need to be perfect, just not bad/broken.  Note that normal rst-style formatting doesn't work in these one-line examples, so I've typically relied on 'single-quotes' rather than ``double-code quotes`` for special terms or code.

* @arezaii:
  - [ ] TOML.chpl
  - [ ] Version.chpl
* @daviditen:
  - [ ] FunctionalOperations.chpl
  - [ ] HDF5.chpl (and  C_HDF5)
  - [ ] NetCDF.chpl
  - [ ] DateTime.chpl
  - [ ] Map.chpl
@dlongnecke-cray:
  - [ ] ConcurrentMap.chpl
  - [ ] List.chpl
  - [ ] Memory.chpl
  - [ ] Memory/*.chpl
  - [ ] Set.chpl
* @e-kayrakli:
  - [ ] SortedMap.chpl
  - [ ] SortedSet.chpl
  - [ ] Heap.chpl
* @lydia-duncan : 
  - [ ] UnitTest.chpl
  - [ ] BigInteger.chpl
  - [ ] Path.chpl 
* @mppf:
  - [ ] AtomicObjects.chpl
  - [ ] Channel.chpl
  - [ ] Crypto.chpl
  - [ ] DistributedBag.chpl
  - [ ] DistributedDeque.chpl
  - [ ] EpochManager.chpl
  - [ ] Futures.chpl
  - [ ] LockFreeQueue.chpl
  - [ ] LockFreeStack.chpl
  - [ ] MPI.chpl (OK that I made sub-module private to remove it from index?)
  - [ ] Socket.chpl
  - [ ] Sort.chpl
  - [ ] Errors.chpl
  - [ ] Reflection.chpl
@ronawho:
  - [ ] CopyAggregation.chpl
  - [ ] PeekPoke.chpl
  - [ ] UnorderedAtomics.chpl
  - [ ] UnorderedCopy.chpl
  - [ ] Barriers.chpl
  - [ ] CommDiagnostics.chpl

File: modules/packages/SortedSet.chpl
Comment (e-kayrakli) [inline modules/packages/SortedSet.chpl:21]: Not broken.

Given that this is right after `SortedMap`, should we strive for consistency between the two?

If so maybe `Provides the 'sortedSet' type for storing sorted unique elements.`?

File: compiler/AST/ModuleSymbol.cpp
Comment (lydia-duncan) [reply compiler/AST/ModuleSymbol.cpp:363]: I think we could do this.  I'd also be tempted to make a specific pragma and apply that, since that will make it less likely we'll need to adjust the compiler again if we miss a case or change our minds or something.

Comment (bradcray) [reply compiler/AST/ModuleSymbol.cpp:358]: I just pushed an improvement to this that looks at the module's filename and squashes it when it starts with "internal/", which feels better though still not ideal.  In the commit message I pointed out that the main downside of this is that it will similarly squash modules passed to `chpldoc` by end users that happen to have a filename of `internal/whatever.chpl`, but it's actually not quite as bad as this—they also have to use `--no-html`.

One other way I could imagine tightening this up would be to pass the full path to the internal modules via the Makefile and to have `chpldoc` see that the modules are in `$CHPL_HOME/modules/internal/`.  I just can't recall how well or poorly the compiler can reason about CHPL_HOME offhand...

I agree that introducing a pragma would be more principled but am a little hesitant to do so given that we haven't needed such a pragma up until now, the number of files it'll touch, and the need to rely on future internal modules to similarly add the pragma.

Comment (lydia-duncan) [reply compiler/AST/ModuleSymbol.cpp:358]: Sounds good!

Comment (bradcray) [reply compiler/AST/ModuleSymbol.cpp:358]: OK, I think the commit I just pushed should do this in an airtight way:  The Makefile passes in the absolute paths to the internal modules we want to document, and the docs.cpp pass looks for files in the same directory and tags them as MOD_INTERNAL.

File: modules/packages/Futures.chpl
Comment (bradcray) [reply modules/packages/Futures.chpl:23]: Ah, now I see.  This is defining a macro-style replacement for `---` for use within the documentation itself.  That's more advanced than my rst/sphinx knew about.

File: compiler/AST/ModuleSymbol.cpp
Comment (lydia-duncan) [reply compiler/AST/ModuleSymbol.cpp:358]: Even better :)

**Comment by arezaii** [comment]
Am I the only one still seeing the internal modules still listed in the module index? (I just downloaded the documentation artifact from the CI check to review it)

**Comment by bradcray** [comment]
Hmm, let me double-check from a clean build...

**Comment by bradcray** [comment]
@arezaii : It looks correct on my desktop, but it could very well be that the approach I used for identifying internal modules isn't portable to the build system (which would suggest potentially not portable to other environments as well).  I may need to push some commits that add more debugging information to understand the build system's configuration better, but will see where I can get without that first.

**Comment by bradcray** [comment]
We resolved the confusion just above—Ahmad was seeing the distribution modules (which were intended to be included but currently have no description) and thinking they were internal.

**Comment by dlongnecke-cray** [comment]
The module blurbs I own look good to me.

**Comment by mppf** [comment]
I checked the boxes - the blurbs you asked me to look at look good to me.

File: modules/standard/CommDiagnostics.chpl
Comment (ronawho) [inline modules/standard/CommDiagnostics.chpl:21]: To me "communication requirements" sounds like the things we require out of the hardware. I might say "Supports counting and reporting network communication operations."

File: modules/packages/SortedSet.chpl
Comment (bradcray) [reply modules/packages/SortedSet.chpl]: Thanks Engin.  I believe I've resolved this in https://github.com/chapel-lang/chapel/pull/19584/commits/e6f587064d1275a38fb0f6d32d1d704ab111c74e

**Comment by bradcray** [comment]
@arezaii : I think at this point, I've got all the fine-grained feedback I wanted on the new module summary lines, and am just looking for you to give the overall PR and code that was not specific to a module's summary line the OK (I know you approved the PR, though that was before all the changes I made since then).  (Or you can defer to someone else if you don't have time or it's exceeded your interests / comfort zone).

**Review by arezaii — APPROVED** [review]
Code changes still look good to me, I checked the CI artifact as well. I did not look at the `.chpl` files, and only briefly looked at the `.rst` files.

**Pull request merged**
